text/0000-dyn-error-generic-member-access.md

```diff
@@ -0,0 +1,527 @@
+    }
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* The `Request` api is being added purely to help with this function. This will
+  likely need some design work to make it more generally applicable, hopefully
+  as a struct in `core::any`.
+* The `context` function name is currently widely used throughout the rust
+  error handling ecosystem in libraries like `anyhow` and `snafu` as an
+  ergonomic version of `map_err`. If we settle on `context` as the final name
+  it will possibly break existing libraries.
+
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## Do Nothing
+
+We could not do this, and continue to add accessor functions to the `Error`
+trait whenever a new type reaches critical levels of popularity in error
+reporting.
+
+If we choose to do nothing we will continue to see hacks around the current
+limitations on the `Error` trait such as the `Fail` trait, which added the
+missing function access methods that didn't previously exist on the `Error`
+trait and type erasure / unnecessary boxing of errors to enable downcasting to
+extract members.
+[[1]](https://docs.rs/tracing-error/0.1.2/src/tracing_error/error.rs.html#269-274).
+
+
+## Use an alternative proposal that relies on the `Any` trait for downcasting
+
+```rust
+pub trait Error {
+    /// Provide an untyped reference to a member whose type matches the provided `TypeId`.
+    ///
+    /// Returns `None` by default, implementors are encouraged to override.
+    fn provide_context(&self, ty: TypeId) -> Option<&dyn Any> {
+        None
+    }
+}
+
+impl dyn Error {
+    /// Retrieve a reference to `T`-typed context from the error if it is available.
+    pub fn context<T: Any>(&self) -> Option<&T> {
+        self.provide_context(TypeId::of::<T>())?.downcast_ref::<T>()
+    }
+}
+```
+
+### Why isn't this the primary proposal?
+
+There are two big issues with using the `Any` trait that I believe justify the
+more complicated solution.
+
+- You cannot return dynamically sized types as `&dyn Any`
+- It's easy to introduce runtime errors with `&dyn Any` by either comparing to
+  or returning the wrong type
+
+By making all the type id comparison internal to the `Request` type it is
+impossible to compare the wrong type ids. By encouraging explicit type
+parameters when calling `provide` the compiler is able to catch errors where
+the type passed in doesn't match the type that was expected. So while the API
+for the main proposal is more complicated it should be less error prone.
+
+# Prior art
+[prior-art]: #prior-art
+
+I do not know of any other languages whose error handling has similar
+facilities for accessing members when reporting errors. For the most part,
+prior art for this proposal comes from within rust itself in the form of
+previous additions to the `Error` trait.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+* What should the names of these functions be?
```

Comment:
I'm most interested in the question raised above about `std::Error::context` conflicting with `failure::Fail::context` and `anyhow::Error::context`. Have you considered a different name for the "reporter-facing" API like `Error::from_context` or `Error::request_context`?

Reply:
I'm not sure I like `from_context`, but `request_context` is a good option. Though I think I'd rather conflict honestly, to encourage downstream error libraries to change their API to match the mental model for the error trait that is proposed here. In those libraries they view the entire chain of errors as one error, where each time you wrap it you're adding context. This isn't consistent with the error trait though, with the API provided here each error is distinct and the saved sources are context. So while its definitely true that errors in a chain have a context-like relationship a bunch of the error handling libraries out there model it backwards and I think that's a source of confusion in the ecosystem and part of why people complain about error handling in rust so much.